File: amethyst.ui/src/types/events/ScoreSheet.ts
import { EventWithBody } from "@/hooks";
import { TeamSide } from "../TeamSide";

export class ScoreSheetJammerNumberSet extends EventWithBody {
    constructor(teamSide: TeamSide, lineNumber: number, jammerNumber: string) {
        super("ScoreSheetJammerNumberSet", { teamSide, totalJamNumber: lineNumber, value: jammerNumber });
    }
}

export class ScoreSheetPivotNumberSet extends EventWithBody {
    constructor(teamSide: TeamSide, totalJamNumber: number, pivotNumber: string) {
        super("ScoreSheetPivotNumberSet", { teamSide, totalJamNumber, value: pivotNumber });
    }
}

export class ScoreSheetLostSet extends EventWithBody {
    constructor(teamSide: TeamSide, lineNumber: number, lost: boolean) {
        super("ScoreSheetLostSet", { teamSide, totalJamNumber: lineNumber, value: lost });
    }
}

export class ScoreSheetLeadSet extends EventWithBody {
    constructor(teamSide: TeamSide, lineNumber: number, lost: boolean) {
        super("ScoreSheetLeadSet", { teamSide, totalJamNumber: lineNumber, value: lost });
    }
}

export class ScoreSheetCalledSet extends EventWithBody {
    constructor(teamSide: TeamSide, lineNumber: number, lost: boolean) {
        super("ScoreSheetCalledSet", { teamSide, totalJamNumber: lineNumber, value: lost });
    }
}

export class ScoreSheetInjurySet extends EventWithBody {
    constructor(lineNumber: number, injury: boolean) {
        super("ScoreSheetInjurySet", { totalJamNumber: lineNumber, value: injury });
    }
}

export class ScoreSheetStarPassTripSet extends EventWithBody {
    constructor(teamSide: TeamSide, jamNumber: number, starPassTrip: number | null) {
        super("ScoreSheetStarPassTripSet", { teamSide, totalJamNumber: jamNumber, starPassTrip })
    }
}

File: jamster.engine/Events/ScoreSheet.cs
using jamster.engine.Domain;
using jamster.engine.Services;

namespace jamster.engine.Events;

public sealed class ScoreSheetTripScoreSet(Guid7 id, ScoreSheetTripScoreSetBody body) : Event<ScoreSheetTripScoreSetBody>(id, body);
public sealed record ScoreSheetTripScoreSetBody(TeamSide TeamSide, int TotalJamNumber, int TripNumber, int? Value) : TeamEventBody(TeamSide);

public sealed class ScoreSheetLeadSet(Guid7 id, ScoreSheetStatsSetBody body) : Event<ScoreSheetStatsSetBody>(id, body);
public sealed class ScoreSheetLostSet(Guid7 id, ScoreSheetStatsSetBody body) : Event<ScoreSheetStatsSetBody>(id, body);
public sealed class ScoreSheetCalledSet(Guid7 id, ScoreSheetStatsSetBody body) : Event<ScoreSheetStatsSetBody>(id, body);
public sealed class ScoreSheetNoInitialSet(Guid7 id, ScoreSheetStatsSetBody body) : Event<ScoreSheetStatsSetBody>(id, body);
public sealed record ScoreSheetStatsSetBody(TeamSide TeamSide, int TotalJamNumber, bool Value) : TeamEventBody(TeamSide);

public sealed class ScoreSheetInjurySet(Guid7 id, ScoreSheetInjurySetBody body) : Event<ScoreSheetInjurySetBody>(id, body);
public sealed record ScoreSheetInjurySetBody(int TotalJamNumber, bool Value);

public sealed class ScoreSheetJammerNumberSet(Guid7 id, ScoreSheetSkaterNumberSetBody body) : Event<ScoreSheetSkaterNumberSetBody>(id, body);
public sealed class ScoreSheetPivotNumberSet(Guid7 id, ScoreSheetSkaterNumberSetBody body) : Event<ScoreSheetSkaterNumberSetBody>(id, body);
public sealed record ScoreSheetSkaterNumberSetBody(TeamSide TeamSide, int TotalJamNumber, string Value) : TeamEventBody(TeamSide);

public sealed class ScoreSheetStarPassTripSet(Guid7 id, ScoreSheetStarPassTripSetBody body) : Event<ScoreSheetStarPassTripSetBody>(id, body);
public sealed record ScoreSheetStarPassTripSetBody(TeamSide TeamSide, int TotalJamNumber, int? StarPassTrip) : TeamEventBody(TeamSide);

public sealed class ScoreSheetLineDeleted(Guid7 id, ScoreSheetLineDeletedBody body) : Event<ScoreSheetLineDeletedBody>(id, body);
public sealed record ScoreSheetLineDeletedBody(int TotalJamNumber);